src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java

```diff
@@ -341,13 +343,22 @@ public void onPaste(PasteEvent event)
 
             final Position start = getSelectionStart();
             
-            Scheduler.get().scheduleDeferred(new ScheduledCommand()
+            Scheduler.get().scheduleFinally(new ScheduledCommand()
```

Comment:
How much time drift can occur between the time the paste event occurs and the time when the scheduler actually runs this? (wondering if we should sample the modifier key state at the time the event occurs rather than later)

Reply:
`scheduleFinally()` basically ensures that the call is run at the end of the current event loop, whereas `scheduleDeferred()` runs it in the next event loop. This actually make sure the call is run before the current edit event completes, and also removes a visual 'flicker' where pasting would briefly show the un-indented text, and then quickly flash to the re-indented text.

Figuring out the active set of keys here is a bit tricky since Ace handles the actual paste event, then generates its own `"paste"` signal with the associated text data. For the reason, we do this dance with the `KeyboardMonitor` to see what modifiers were actually held during the time of paste. (In particular, the browser doesn't report the state of modifier keys as part of the paste event, so we need to explicitly track keydown / keyup events that occur to figure out what modifiers are actually down at the time of paste)

Reply:
Yes, it feels a little weird that we have to be running code constantly in the background to get this right, but since Ace's paste signal doesn't include the originating event I realize it's hard to do better!